sw/device/tests/crypto/cryptotest/firmware/hash.c

```diff
@@ -0,0 +1,150 @@
+const unsigned int kOtcryptoHashDigestWordsSha384 = 12;
+const unsigned int kOtcryptoHashDigestWordsSha512 = 16;
+const unsigned int kOtcryptoHashDigestWordsSha3_256 = 8;
+const unsigned int kOtcryptoHashDigestWordsSha3_384 = 12;
+const unsigned int kOtcryptoHashDigestWordsSha3_512 = 16;
+
+status_t handle_hash(ujson_t *uj) {
+  // Declare test arguments
+  cryptotest_hash_algorithm_t uj_algorithm;
+  cryptotest_hash_shake_digest_length_t uj_shake_digest_length;
+  cryptotest_hash_message_t uj_message;
+  // Deserialize test arguments from UART
+  TRY(ujson_deserialize_cryptotest_hash_algorithm_t(uj, &uj_algorithm));
+  TRY(ujson_deserialize_cryptotest_hash_shake_digest_length_t(
+      uj, &uj_shake_digest_length));
+  TRY(ujson_deserialize_cryptotest_hash_message_t(uj, &uj_message));
+
+  // Create input message
+  uint8_t msg_buf[MaxMessageBytes];
+  memcpy(msg_buf, uj_message.message, uj_message.message_len);
+  otcrypto_const_byte_buf_t input_message = {
+      .len = uj_message.message_len,
+      .data = msg_buf,
+  };
+
+  // Create digest skeleton
+  uint32_t digest_buf[MaxDigestWords];
+  memset(digest_buf, 0, MaxDigestWords * sizeof(uint32_t));
+  otcrypto_hash_digest_t digest = {
+      .data = digest_buf,
+  };
+
+  // Handle to correct oneshot hash API for the provided algorithm
+  otcrypto_status_t (*hash_oneshot)(otcrypto_const_byte_buf_t,
+                                    otcrypto_hash_digest_t);
+
+  uint8_t test_stepwise = false;
+  otcrypto_hash_mode_t mode;
+  switch (uj_algorithm) {
+    case kCryptotestHashAlgorithmSha256:
+      mode = kOtcryptoHashModeSha256;
+      digest.len = kOtcryptoHashDigestWordsSha256;
+      hash_oneshot = otcrypto_hash;
+      test_stepwise = true;
+      break;
+    case kCryptotestHashAlgorithmSha384:
+      mode = kOtcryptoHashModeSha384;
+      digest.len = kOtcryptoHashDigestWordsSha384;
+      hash_oneshot = otcrypto_hash;
+      test_stepwise = true;
+      break;
+    case kCryptotestHashAlgorithmSha512:
+      mode = kOtcryptoHashModeSha512;
+      digest.len = kOtcryptoHashDigestWordsSha512;
+      hash_oneshot = otcrypto_hash;
+      test_stepwise = true;
+      break;
+    case kCryptotestHashAlgorithmSha3_256:
+      mode = kOtcryptoHashModeSha3_256;
+      digest.len = kOtcryptoHashDigestWordsSha3_256;
+      hash_oneshot = otcrypto_hash;
+      break;
+    case kCryptotestHashAlgorithmSha3_384:
+      mode = kOtcryptoHashModeSha3_384;
+      digest.len = kOtcryptoHashDigestWordsSha3_384;
+      hash_oneshot = otcrypto_hash;
+      break;
+    case kCryptotestHashAlgorithmSha3_512:
+      mode = kOtcryptoHashModeSha3_512;
+      digest.len = kOtcryptoHashDigestWordsSha3_512;
+      hash_oneshot = otcrypto_hash;
+      break;
+    case kCryptotestHashAlgorithmShake128:
+      mode = kOtcryptoHashXofModeShake128;
+      // length in words = ceil(length in bytes / word size)
+      digest.len = (uj_shake_digest_length.length + 3) / sizeof(uint32_t);
```

Comment:
There exists a [`ceil_div`](https://github.com/lowRISC/opentitan/blob/6f251215344949e9dcc596c67860defcba856d34/sw/device/lib/base/math.h#L60) implementation in math.h.